rudder-core/src/main/scala/com/normation/rudder/datasources/DataSourceRepository.scala

```diff
@@ -48,4 +68,212 @@ trait DataSourceRepository {
   def save(source : DataSource) : Box[DataSource]
 
   def delete(id : DataSourceId) : Box[DataSource]
-}
\ No newline at end of file
+}
+
+/*
+ * A trait that exposes interactive callbacks for
+ * data sources, i.e the method to call when one
+ * need to update datasources.
+ */
+trait DataSourceUpdateCallbacks {
+
+  def onNewNode(node: NodeId): Unit
+  def onGenerationStarted(generationTimeStamp: DateTime): Unit
+  def onUserAskUpdateAll(actor: EventActor): Unit
+  def onUserAskUpdateNode(actor: EventActor, nodeId: NodeId): Unit
+
+  /*
+   * Initialise all datasource so that they are ready to schedule their
+   * first data fetch or wait for other callbacks.
+   *
+   * Non periodic data source won't be updated with that call.
+   * Periodic one will be updated in a random interval between
+   * 1 minute and min(period / 2, 30 minute) to avoid to extenghish
+   * all resources on them.
+   */
+  def startAll(): Unit
+}
+
+class MemoryDataSourceRepository extends DataSourceRepository {
+
+  private[this] var sources : Map[DataSourceId,DataSource] = Map()
+
+  def getAll() = synchronized(Full(sources))
+
+  def get(id : DataSourceId) : Box[Option[DataSource]]= synchronized(Full(sources.get(id)))
+
+  def save(source : DataSource) = synchronized {
+    sources = sources +  ((source.id,source))
+    Full(source)
+  }
+
+  def delete(id : DataSourceId) : Box[DataSource] = synchronized {
+    sources.get(id) match {
+      case Some(source) =>
+        sources = sources - (id)
+        Full(source)
+      case None =>
+        Failure(s"Data source '${id}' does not exists, and thus can't be deleted")
+    }
+  }
+}
+/**
+ * This is the higher level repository facade that is managine the "live"
+ * instance of datasources, with the scheduling initialisation and update
+ * on different repository action.
+ *
+ * It doesn't deal at all with the serialisation / deserialisation of data source
+ * in data base.
+ */
+class DataSourceRepoImpl(
+    backend: DataSourceRepository
+  , fetch  : QueryDataSourceService
+  , uuidGen: StringUuidGenerator
+) extends DataSourceRepository with DataSourceUpdateCallbacks {
+
+  private[this] var datasources = Map[DataSourceId, DataSourceScheduler]()
```

Comment:
Can it be considered as a cache ? (and get informations for getAll / get in instead of going directly to backend)